### ISSUE TITLE
Develop less-frequent "comprehensive tests"

### ISSUE THREAD

**Issue by maxachis** [Police-Data-Accessibility-Project/data-sources-app#412]
### Context

* As the number of tests we have expands, it becomes less practical to run all of them (particularly the edge cases) on every PR into dev. 
* Many of these tests are error-catching tests which would be captured in various unit tests, but which should ideally be run on every endpoint which uses them, to make sure there are no surprises when they're all tied together.
* Ideally, these would be easy to design and deploy en-masse for various endpoints. But they would be time-intensive.

### Requirements

* Separate some tests into tests which are run on every pull into main, rather than every PR period. 
* These tests would test the less-frequent branches of logic end-to-end (for example, bad parameters, a database failure, authentication failures). 

### Tests

* Plenty of them! 

### Docs

* Will need to develop documentation explaining and differentiating these from other tests.

### Open questions

* Any uncertainties? Anything we're flexible about?

**Comment by josh-chamberlain** [comment]
Nice! Is there a simple way we can delineate in the code which `destination` branch each test is run on? Seems like most tests would be `dev` and some would be `dev`, `main`.

Would this require _new_ tests, or can we establish this concept with the existing suite and treat further test creation as its own thing?

**Comment by maxachis** [comment]
> Nice! Is there a simple way we can delineate in the code which `destination` branch each test is run on? Seems like most tests would be `dev` and some would be `dev`, `main`.
> 

Yes! We can either update our existing Github Action or create a separate GitHub action to these more comprehensive tests, which ensures they run on PRs into main.

> Would this require _new_ tests, or can we establish this concept with the existing suite and treat further test creation as its own thing?

We could establish this with the existing suite, although most of the tests we have so far do one of the following:
* End-to-end test the happy path of existing endpoints
* Unit test existing functionality (which take very little time)
* Test Database Client methods 

All of these I would argue are things we should be testing on every PR, and combined don't take more than 2 minutes. 

The new tests I'm envisioning would do a lot more testing, but with the caveat that they'll be testing the same functionality across all the different endpoints, and hence would involve less logic -- it's just checking to see that all the endpoints have the same, consistent behavior. 

An example for this is my `test_endpoints.py` file, which tests that only the HTTP methods we intend for each endpoint return a response, and that others return a "Method Not Allowed" error.  It's looking for the same response each time -- it's just testing for consistency.

**Comment by maxachis** [comment]
A further update to all of this -- as I've refined my testing schema, integration tests have slowly become more comprehensive. In addition to the happy path of tests, tests often include other basic tests -- for example, that calls with certain headers yield unauthorized responses. 

Consequently, the test suite is _more_ comprehensive, but is not _comprehensive_. 

Full comprehensiveness is probably a pipe dream. But there are some test patterns we could probably deploy fairly consistently simply to check proper behavior. 

Requirements

- [ ] Create `tests-comprehensive` directory. All further tests described here will be placed in that
- [ ] `test-api-key`: All endpoints where an API key is _not_ listed as an option in the documentation should produce a `BAD REQUEST` result.
- [ ] `test-jwt`: All endpoints where a JWT (aka, "Bearer" auth) is _not_ listed as an option in the documentation should produce a `BAD REQUEST` result.
- [ ] `test-expired`: All endpoints which use a JWT, but where the JWT is expired, should produce an `UNAUTHORIZED` result
- [ ] `test-csrf`: All endpoints which use a JWT, but where the JWT in question is encrypted using a key different than the one present on the app, should produce an `UNAUTHORIZED` result

**Comment by josh-chamberlain** [comment]
@maxachis these are certainly comprehensive! Maybe bordering on overkill/diminishing returns. Some ideas for streamlining from our mutual frenemy, which I have also streamlined: 

Focus on Representative Cases
- Test a few endpoints per category (e.g., one for API keys, one for JWTs) rather than duplicating the tests for every endpoint.

Centralized Enforcement:
- If headers and JWTs are validated in a middleware or decorator, focus tests on the middleware logic and ensure all endpoints use it correctly.

Focus tests on high-risk scenarios:
- Endpoints that modify data or handle sensitive information.
- Public-facing endpoints or those without authentication.
- Areas with complex access control logic.

**Comment by maxachis** [comment]
@josh-chamberlain Agreed about the bordering on overkill! I will say, however, that it's somewhat less of a border because of an ingenious (if I do say so myself) method of parsing the API specification (which can be accessed from https://data-sources-v2.pdap.dev/api/swagger.json) and then running tests on endpoints based on what qualities they have.

For example, a test that retrieves all endpoints which list an authorization header in the documentation and checks to see that they all list `400: BAD REQUEST` as their possible responses. And then another test which takes those endpoints and checks to see that they produce `400: BAD REQUEST` under appropriate conditions. 

This helps save time, but also helps ensure consistency between documentation and internal logic. And indeed, these tests prompted me to refactor some existing code in order to better ensure that consistency.

That being said -- yes, there is diminishing returns to this (one of the reasons I put this lowest on my pre-v1-sunset playlist), and I do not believe this has to be done in its entirety on any particular timetable -- rather, this is a means to fill in cracks where we've time.

**Comment by josh-chamberlain** [comment]
@maxachis wow, that _is_ rather ingenious!
